main/src/ca/uwaterloo/flix/language/ast/Purity.scala

```diff
@@ -80,11 +104,64 @@ object Purity {
 
   /**
     * Returns the purity of the given formula `eff`. Returns [[Pure]] for the
-    * effect constant [[TypeConstructor.Pure]] and [[Impure]] otherwise.
+    * effect constant of [[TypeConstructor.Pure]], returns [[Impure]] for the
+    * effect constant of [[Symbol.IO]], and [[ControlImpure]] otherwise.
+    *
+    * Assumes that the given type is a well-formed formula without variables,
+    * aliases, or associated types.
+    */
+  def fromType(eff: Type): Purity = {
+    evaluateFormula(eff) match {
+      case set if set.isEmpty => Purity.Pure
+      case set if set.hasSize(1) && set.contains(Symbol.IO) => Purity.Impure
```

Comment:
I would worry that this check is not always complete.

Reply:
Yeah, but it is wrong to the safe side. Additionally - given the conjecture that no actual effects will be infinite (even with an infinite effect domain) - it is accurate since evaluateFormula will never return `Compl`